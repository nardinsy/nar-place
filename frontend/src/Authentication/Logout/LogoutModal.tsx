import { useContext } from "react";
import { useHistory } from "react-router-dom";
import MessageModal from "../../Shared-UI/MessageModal";
import AuthContext from "../../store/auth-context";
import Button from "../../Shared-UI/Button";

const LogoutModal = (props) => {
  const history = useHistory();
  const authContext = useContext(AuthContext);

  const logoutClient = async (event) => {
    event.preventDefault();
    event.stopPropagation();
    await authContext.logout();
    history.replace("/");
  };

  const cancelLogoutHandler = (event) => {
    event.preventDefault();
    event.stopPropagation();
    history.replace("/");
  };

  return (
    <MessageModal message={"Do you want to log out?"}>
      <Button onClick={logoutClient} action={"submit"}>
        Logout
      </Button>
      <Button onClick={cancelLogoutHandler} action={"cancel"}>
        Cancel
      </Button>
    </MessageModal>
  );
};

export default LogoutModal;
