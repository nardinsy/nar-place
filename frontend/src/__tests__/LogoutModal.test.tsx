import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import LogoutModal from "../Authentication/Logout/LogoutModal";
import AuthContext, { AuthContextT } from "../contexts/auth-context";
import { BrowserRouter } from "react-router-dom";
import ReactDOM from "react-dom";
import { ReactNode, ReactPortal } from "react";

const mockHistoryReplace = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    replace: mockHistoryReplace,
  }),
}));

const renderLogoutModal = () => {
  const authProviderValue = {
    isLoggedin: true,
    token: "1234",
    username: "nardin",
    userPictureUrl: undefined,
    userId: "1234",
    logout: jest.fn(() => Promise.resolve()),
    setPictureUrl: jest.fn(),
    setUsername: jest.fn(),
  } satisfies AuthContextT;

  const { baseElement } = render(
    <BrowserRouter>
      <AuthContext.Provider value={authProviderValue}>
        <LogoutModal />
      </AuthContext.Provider>
    </BrowserRouter>
  );

  return { baseElement, ...authProviderValue };
};

describe("Logout modal", () => {
  const oldCreatePortal = ReactDOM.createPortal;
  beforeAll(() => {
    ReactDOM.createPortal = (node: ReactNode): ReactPortal =>
      node as ReactPortal;
  });

  afterAll(() => {
    ReactDOM.createPortal = oldCreatePortal;
  });

  test("logs out correctly", async () => {
    const { logout, baseElement } = renderLogoutModal();

    const logoutButton = screen.getByRole("button", {
      name: /logout/i,
    });
    fireEvent.click(logoutButton);

    expect(baseElement).toMatchSnapshot();
    expect(logoutButton).toBeInTheDocument();

    // await expect(logout).toHaveBeenCalledTimes(1);
    await waitFor(() => expect(logout).toHaveBeenCalledTimes(1));

    expect(mockHistoryReplace).toHaveBeenCalledTimes(1);
    expect(mockHistoryReplace).toHaveBeenCalledWith("/");
  });

  test("cancels correctly", () => {
    const { baseElement } = renderLogoutModal();

    const cancelButton = screen.getByRole("button", {
      name: /cancel/i,
    });
    fireEvent.click(cancelButton);

    expect(baseElement).toMatchSnapshot();
    expect(cancelButton).toBeInTheDocument();

    expect(mockHistoryReplace).toHaveBeenCalledTimes(1);
    expect(mockHistoryReplace).toHaveBeenCalledWith("/");
  });
});
