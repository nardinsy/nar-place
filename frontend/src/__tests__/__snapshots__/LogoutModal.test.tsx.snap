// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout modal cancel correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal cancels correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal logout correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal logs out correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal navigates to home page 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on cancel button click cancels correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on cancel button click navigates to home page 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on log out button click logs out correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on log out button click navigates to home page 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on log out button click on cancel button click cancels correctly 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Logout modal on log out button click on cancel button click navigates to home page 1`] = `
<body>
  <div>
    <div
      class="backdrop"
    />
    <div
      class="modal undefined"
    >
      <div
        class="content"
      >
        <div
          class="message-container"
        >
          <div
            class="message"
          >
            Do you want to log out?
          </div>
          <div>
            <button
              class="button undefined button-submit"
            >
              Logout
            </button>
            <button
              class="button undefined button-cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
